### PR TITLE
Feature/model functions

### PR DESCRIPTION
**Pull request by paddy7wb**
Functionalised code from modelling part of `mrc_ide/circumcision_coverage` 
analysis, with appropriate documentation. (Somewhat lazily) Tested by rewritting 
modelling part of analysis to incorporate functions, running for LSO, and
comparing with original model output, which is saved on Sharepoint at 
https://imperiallondon.sharepoint.com/:u:/r/sites/HIVInferenceGroup-WP/Shared%20Documents/Circumcision%20coverage/Runs/LSO_Results_DistrictAgeTime_ByType.csv.gz?csf=1&web=1&e=zvENNe
and 
https://imperiallondon.sharepoint.com/:u:/r/sites/HIVInferenceGroup-WP/Shared%20Documents/Circumcision%20coverage/Runs/LSO_TMBObjects_DistrictAgeTime_ByType.RData?csf=1&web=1&e=8vOjXe,
found to be the same.

Still to do on this branch/section:
1. Write more formal, smaller reference tests for functions. Pull original fits for LSO
from sharepoint, too large to store in package
2. Rewrite parts of "create_hazard" and "create_integration" matrix functions
so common code is in unexported functions.
3. Perhaps create master function to perform all "matrix creation" steps up to 
and included creating `dat_tmb` object.
5. Check with Jeff to see if "layout" of functionalised behaviour from modelling
is correct (i.e. one step to prepare survey data, one to create TMB model data,
etc.).
6. Need to test if "imports" are functioning properly, particularly those from
`threemc` itself.

### CODE REVIEW

File: .lintr [hunk 1]
```diff
@@ -2,6 +2,8 @@ linters: with_defaults(
     object_length_linter = NULL,
     object_usage_linter = NULL,
     todo_comment_linter = NULL,
-    cyclocomp_linter = NULL
+    cyclocomp_linter = NULL,
+    trailing_whitespace_linter = NULL,
```
Comment (r-ash) [inline .lintr:6]: You can make RStudio remove this for you automatically on save. In tools -> global options -> code -> saving -> strip trailing horizontal whitespace when saving

File: R/circ_sample_tmb.R [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+circ_sample_tmb <- function(obj, opt, n_samples = 1000, ...) {
+  
+  # Getting the TMB into "Naomi" format to sample from using the NAOMI package
+  opt$par.fixed <- opt$par
+  opt$par.full <- obj$env$last.par
+  fit <- c(opt, obj = list(obj))
+  class(fit) <- "naomi_fit"
+
+  # Look at standard deviation report
+  fit$sdreport <- sdreport(fit$obj, fit$par, getJointPrecision = TRUE)
```
Comment (r-ash) [inline R/circ_sample_tmb.R:33]: I would favour qualifying your function calls over using `importFrom` i.e. use `TMB::sdreport(...)`

File: R/compile_tmb.R [hunk 3]
```diff
@@ -0,0 +1,40 @@
+# wrapper around TMB::compile which silences annoying C++ warning messages
+#' @title Compile TMB model without C++ Warnings
+#' @description Wrapper around \code{\link[TMB]{compile}} which optionally 
+#' redirects C++ compiler output to a logfile, avoiding your R console, since 
+#' it can be quite verbose. Note that \code{logfile} argument default is not
+#' valid for Windows.
```
Comment (r-ash) [inline R/compile_tmb.R:6]: Up to you but you but you can get away without specifying this explicitly. Can do
```
#' This is the title
#'
#' This is my description whihc might be
#' multiple lines long
#'
#' @param ...
```

File: R/compile_tmb.R [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+#' @param file C++ file.
+#' @param logfile File to redirect output from C++ compiler. To print output 
+#' to R console, specify \code{logfile = NULL}, Default: '/tmp/tmb_logfile.log'
+#' @param ... Additional parameters for \code{\link[TMB]{compile}}.
+#' 
+#' @seealso 
+#'  \code{\link[TMB]{compile}}
+#' @rdname compile_tmb
```
Comment (r-ash) [inline R/compile_tmb.R:14]: You can avoid this too when the rdname is the same as the function name

File: R/compute_quantiles.R [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+# function to calculate quantiles for rates and cumulative hazard
+# maybe add "textProgressBar" to for loop if it's too slow?
+compute_quantiles <- function(out, 
+                              fit, 
+                              probs = c(0.5, 0.025, 0.975), 
+                              names = FALSE, ...) {
+  
+  probs <- sort(probs)
+  
+  # function to do "legwork" of computing quantiles
```
Comment (r-ash) [inline R/compute_quantiles.R:37]: Is Jeff going to work on this package? If so he will probably ask you to use double ##, I think single # means something special to emacs which causes it to format these lines weirdly

File: R/create_integration_matrices.R [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+#' 
+#' @seealso
+#'  \code{\link[threemc]{create_shell_dataset}}
+#'  \code{\link[threemc]{create_integration_matrix_agetime}}
+#'  \code{\link[threemc]{create_integration_matrix_agetime_lag}}
+#' @rdname create_integration_matrices
+#' @export
+#' 
+#' @importFrom threemc create_integration_matrix_agetime 
+#' @importFrom threemc create_integration_matrix_agetime_lag
```
Comment (r-ash) [inline R/create_integration_matrices.R:31]: You shouldn't need these imports from your own package

File: R/create_integration_matrices.R [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                                        ...
+                                        ) {
+  
+  out <- out
+  out$time1 <- out$time - out$circ_age
+  out$time2 <- out$time
+  out$age <- out$circ_age + 1
+  
+  # Matrix for selecting instantaneous hazard rate
+  IntMat1 <- create_integration_matrix_agetime(
```
Comment (r-ash) [inline R/create_integration_matrices.R:47]: I'd try and stick to snake_case naming convention. Having a standard naming case is supposed to make it easier to read and write code in the package as style is predictable

File: R/prepare_survey_data.R [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+  tmp <- with(survey_circumcision, as.data.frame(table(survey_id, type))) %>% 
+    group_by(survey_id) %>%
+    # calculate percentage and find surveys with all missing data
+    mutate(Freq = Freq / sum(Freq)) %>%
+    filter(type == "Missing", Freq == 1)
+  
+  # return warning detailing all surveys which are missing
+  if (nrow(tmp) > 1) {
+    for (i in seq_len(nrow(tmp))) {
+      warning(paste0(tmp$survey_id[i], " has all type == \"Missing\", and will
```
Comment (r-ash) [inline R/prepare_survey_data.R:174]: Up to you but Rich usually avoids warnings, see blog post here https://reside-ic.github.io/blog/a-warning-about-warning/

File: R/create_integration_matrix_agetime_lag.R [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                       (as.numeric(x["time2_cap2"]) - 1) <= 
+                       timecaps[2] - timecaps[1])))
+      }
+      test <- test[-length(test)]
+      return(test)
+    })
+    cols <- unlist(cols)
+    
+    # Row entries for integration matrix
+    rows <- apply(dat, 1, function(x) {
```
Comment (r-ash) [inline R/create_integration_matrix_agetime_lag.R:125]: apply will automatically convert types for you in the output. Might be fine in this case, I can see you have some explicit casts to as.numeric below. But can sometimes be a cause of unexpected bugs. I would usually think about looping with 

```
lapply(nrow(dat), function(i) {
  row <- dat[i, ]
  ## Do stuff with the row
})

File: R/read_circ_data.R [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+#' @export
+#'
+#' @import dplyr
+#' @importFrom data.table fread
+#' @importFrom sf read_sf
+#' @importFrom rlang sym
+  
+# maybe add a warning for missing "circ" columns for surveys?? And add them in
+# NAs in this situation (look at KEN for this)
+read_circ_data <- function(path, filters = NULL) {
```
Comment (r-ash) [inline R/read_circ_data.R:26]: Does this still link the docs to the function when there is a gap and a comment in between the docs and function declaration? Probably something to check. If you install your package and do `?read_circ_data` do you get the right documentation with it?

File: R/read_circ_data.R [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  
+  # read in data, depending on file type
+  if (grepl(".geojson", path)) {
```
Comment (r-ash) [inline R/read_circ_data.R:29]: You could also use `tools::file_ext` to get this

File: R/read_circ_data.R [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    cols <- names(filters)
+    vals <- as.vector(filters[seq_along(filters)])
+    for (i in seq_along(filters)) {
+      if (!cols[i] %in% names(.data)) next
+      # change col i to symbol (if present), evaluate corresponding filter
+      .data <- filter(.data, !!sym(cols[i]) == vals[i])
+    }
+  }
+  # for areas, add unique identifier within Admin code and merge to boundaries
+  if ("sf" %in% class(.data)) {
```
Comment (r-ash) [inline R/read_circ_data.R:44]: You can use `inherits` to check classes
```
if (inherits(.data, "sf")) {
```

**Review by r-ash — APPROVED** [review]
This looks really good, very tidy code and nice separation into functions. Few suggestions but largely up to you if you want to take them or not. Will approve to not block you over Christmas & new year

File: R/compute_quantiles.R (hunk 5 above)
Comment (jeffeaton) [reply R/compute_quantiles.R:37]: Yeah, Emacs Speaks Statistics default indenting centers lines that are commented with a single `#`. I'm sure there is a way to change that, but I'm not clever enough to figure out how.

File: R/read_circ_data.R (hunk 11 above)
Comment (paddy7wb) [reply R/read_circ_data.R:27]: Yep, just changed that, thanks!

File: R/compute_quantiles.R (hunk 5 above)
Comment (paddy7wb) [reply R/compute_quantiles.R]: I think you can do something like `(setq ess-indent-with-fancy-comments nil)` in your emacs config Jeff, but I've changed all the comments on their own lines to have (at least) two #s

File: R/circ_sample_tmb.R (hunk 2 above)
Comment (paddy7wb) [reply R/circ_sample_tmb.R]: Thanks, have that all changed and will going forward will specify the namespace, agreed it's more clear, I was just getting excited with all the `ryoxygen2` tags lol

File: R/create_integration_matrices.R (hunk 7 above)
Comment (paddy7wb) [reply R/create_integration_matrices.R]: That was a legacy from Matt's code that I never even spotted, changed accordingly

File: R/prepare_survey_data.R (hunk 8 above)
Comment (paddy7wb) [reply R/prepare_survey_data.R:173]: Changed to a message, thanks for that. Very interesting blog, always thought the warning from `install.packages` for an unknown package name was ridiculous alright

File: R/read_circ_data.R (hunk 10 above)
Comment (paddy7wb) [reply R/read_circ_data.R]: Going to have to merge pull request first before looking at this, but I'll check

File: R/create_integration_matrix_agetime_lag.R (hunk 9 above)
Comment (paddy7wb) [reply R/create_integration_matrix_agetime_lag.R]: Went with Jeff's solution

**Pull request merged**